### ISSUE TITLE
Homo oligomer design?

### ISSUE THREAD

**Issue by JSATacoTruck** [richardshuai/fampnn#15]
Hi Richard,

A lab mate of mine told recently told me about some of your groups work, so I started reading into fampnn. The side chain conditioning is very nice, this is very cool work. I wanted to ask you, is there any way to do a homo oligomer design, as in enforce the same sequence for 2+ conformational states? 

looking forward to hearing back from you.

**Comment by richardshuai** [comment]
Hi, I’m glad you liked our work! Unfortunately, we don’t have a builtin way to do homo oligomer design right now, but I imagine you can do something similar to ProteinMPNN’s tied sampling approach by averaging logits across the two chains as you sample. But this is a good feature to have, and I’ll keep this in mind to add when I get the chance.

**Comment by JSATacoTruck** [comment]
awesome! thanks for letting me know. Will keep my eyes peeled for it

**Comment by TienMPhan** [comment]
Hi Richard,

Very nice work! I really enjoyed testing FAMPNN. Just follow up with the issue; do you have a chance to look at the homo oligomer design?

**Comment by richardshuai** [comment]
Hi Tien thanks for following up! Our lab has recently moved over to Caliby for structure-conditioned sequence design (https://github.com/ProteinDesignLab/caliby), which we have found to design much better sequences overall. If you're happy with switching over, I can implement this in Caliby for you?

**Comment by TienMPhan** [comment]
Hi Richard,  
Based on your suggestion on averaging the logits, I also jot down some ideas to implement it.

Here’s the core concept :

- Homo-oligomer: A protein with multiple identical chains (e.g., a homodimer has two identical chains, A and B).

- Symmetry Requirement: For the chains to be identical, the amino acid at a given position on one chain (e.g., residue 10 on chain A) must be the same as the amino acid at the equivalent position on the other chains (e.g., residue 10 on chain B).

- Logits: During sampling, the model predicts "logits" for each position. A logit is a raw, unnormalized score for each of the 20 possible amino acids. These scores are then converted into probabilities (usually via a softmax function) from which a single amino acid is sampled.

- "Averaging Logits" (Tied Sampling): To enforce symmetry, you intervene in this process:

1) at each sampling step, let the model predict logits for all residues in all chains.

2) identify the symmetrically equivalent positions (e.g., A10 and B10).

3) take the logit vector for A10 and the logit vector for B10 and compute their average.

4) replace the original logits for both A10 and B10 with this new, averaged logit vector.

5) then convert this averaged logit vector into probabilities and sample once.

6) Because both positions now share the exact same probability distribution, the amino acid sampled will be assigned to both A10 and B10, guaranteeing identity.

But, it would be great if you could implement it. I'm happy to switch over.

Thanks again for a very nice package!

**Comment by richardshuai** [comment]
Yes this scheme looks right to me! I'm out of town until the 17th, but when I get back I'll do something similar for Caliby and let you know

**Comment by richardshuai** [comment]
@TienMPhan  and @JSATacoTruck, this is now implemented in the Caliby repo (https://github.com/ProteinDesignLab/caliby). Thanks for your patience, and please let me know if you run into any issues!

**Comment by JSATacoTruck** [comment]
awesome! thanks for the heads up!
